singleagent/contrastive_dyna.py

```diff
@@ -0,0 +1,435 @@
+            return scale_gradient(out, self.scale_grad)
+
+        # Check if we are dealing with batched input
+        if action_onehot.ndim == 2:
+            # Use jax.vmap for vectorized/batched operation
+            return jax.vmap(apply_transition)(action_onehot, state)
+        else:
+            # Apply transition directly for unbatched/single input
+            return apply_transition(action_onehot, state)
+        
+        
+#! old codebase did from acme.jax.networks import duelling
+# not sure if this still works
+# created this in case, assuming that it follows the typical dueling 
+# network architecture where it splits the network towards the end 
+# into two streams (one for state value and one for action advantages), 
+# and then recombines them to produce Q-values
+class DuelingMLP(nn.Module):
+    num_actions: int
+    hidden_sizes: list
+
+    @nn.compact
+    def __call__(self, x: jax.Array) -> jax.Array:
+        for size in self.hidden_sizes:
+            x = nn.Dense(features=size)(x)
+            x = nn.relu(x)
+
+        # Separate streams for state value and action advantages
+        value = nn.Dense(features=1)(x)
+        advantages = nn.Dense(features=self.num_actions)(x)
+
+        # Combine streams: Q = V(s) + A(s, a) - mean(A(s, a))
+        q_values = value + advantages - jax.numpy.mean(advantages, axis=1, keepdims=True)
+        return q_values
+    
+#! modified from old codebase prediction_fn
+class PredictionModule(nn.Module):
+    num_actions: int
+    q_dim: int  # old codebase got this from config, not sure where to get this
+
+    @nn.compact
+    def __call__(self, state: jax.Array) -> Any:
+        # Q-values from Dueling MLP
+        # in old codebase it was 
+        # q_values = duelling.DuellingMLP(
+        # num_actions, hidden_sizes=[config.q_dim])(state)
+        q_values = DuelingMLP(num_actions=self.num_actions, hidden_sizes=[self.q_dim])(state)
+
+        # Rewards from a simple MLP, old codebase used
+        # rewards = hk.nets.MLP([config.q_dim, 1])(state)
+        rewards = nn.Dense(features=self.q_dim)(state)
+        rewards = nn.relu(rewards)  # Assuming a ReLU activation before the final layer
+        rewards = nn.Dense(features=1)(rewards)
+
+        rewards = jax.numpy.squeeze(rewards, axis=-1)
+
+        # I copied the Prediction structure from contrastiveDyna and put it on the top
+        return Predictions(
+            state=state, 
+            rewards=rewards, 
+            q_values=q_values)
+
+#！
+class AgentDyna(nn.Module):
+    action_dim: int
+    hidden_dim: int
+    init_scale: float
+    num_actions: int  # Number of possible actions
+    num_blocks: int  # Number of blocks in SimpleTransition
+    scale_grad: float  # Gradient scaling factor
+    q_dim: int # old codebase got this from config, not sure where to get this
+
+    def setup(self):
+
+        #? not sure if we need this
+        self.rnn = vbb.ScannedRNN(
```

Comment:
haha so normally you would, but for these very simple environments, DummyRNN is fine, you probably want the rnn to be an argument where you can choose one of these as options.